File: src/sys/sdl3/NativeEventDisposition.cpp
// Copyright (C) 2026 DarkMatter Productions. GPL-3.0-or-later.
#include "NativeEventDisposition.h"
#include <atomic>
#include <limits>

namespace openq4 {
namespace {
constexpr std::array<NativeDispositionPass,4> SessionFirstOrder{
    NativeDispositionPass::SessionInitial, NativeDispositionPass::MousePoll,
    NativeDispositionPass::KeyboardPoll, NativeDispositionPass::SessionDeferred};
constexpr std::array<NativeDispositionPass,4> PollFirstOrder{
    NativeDispositionPass::MousePoll, NativeDispositionPass::KeyboardPoll,
    NativeDispositionPass::SessionInitial, NativeDispositionPass::SessionDeferred};
std::atomic<std::uint64_t> nextLedger{1};
std::uint64_t Identity() noexcept {
    auto value = nextLedger.load();
    while (value && !nextLedger.compare_exchange_weak(value,
        value == (std::numeric_limits<std::uint64_t>::max)() ? 0 : value + 1)) {}
    return value;
}
void Error(std::string& error, const char* message) noexcept {
    try { error = message; } catch (...) { error.clear(); }
}
}
struct NativeEventDispositionLedger::Guard {
    NativeEventDispositionLedger& ledger;
    ~Guard() { ledger.calling = false; }
};
NativeEventDispositionLedger::NativeEventDispositionLedger(NativeDispositionProbe& value)
    : thread(std::this_thread::get_id()), probe(value), identity(Identity()), records(0), emissions(0) {}
bool NativeEventDispositionLedger::Fail(std::string& error, const char* message) noexcept {
    phase = Phase::Retired;
    Error(error, message);
    return false;
}
bool NativeEventDispositionLedger::Enter(std::string& error) noexcept {
    if (thread != std::this_thread::get_id()) {
        Error(error, "Native disposition is owned by another thread"); return false;
    }
    if (calling) return Fail(error, "Reentrant native disposition requires retirement");
    if (phase == Phase::Retired) { Error(error, "Native disposition requires retirement"); return false; }
    calling = true; return true;
}
bool NativeEventDispositionLedger::Check(std::string& error) noexcept {
    try {
        if (phase == Phase::Retired || !probe.Current(context) || !ingress || !source ||
            !ingress->Validate(*source, receipt.batch, error) || phase == Phase::Retired ||
            !probe.Current(context)) return Fail(error, "Native disposition continuity or provider changed");
        return true;
    } catch (...) { return Fail(error, "Native disposition source failed"); }
}
bool NativeEventDispositionLedger::Begin(NativeQueueIngress& input, NativeQueueSource& provider,
    const NativeQueueBatch& batch, std::string& error) noexcept {
    return BeginInternal(input, provider, batch, false, NativeDispositionSchedule::BeforeSessionDrain, error);
}
bool NativeEventDispositionLedger::BeginPlanned(NativeQueueIngress& input, NativeQueueSource& provider,
    const NativeQueueBatch& batch, std::string& error) noexcept {
    return BeginPlanned(input, provider, batch, NativeDispositionSchedule::BeforeSessionDrain, error);
}
bool NativeEventDispositionLedger::BeginPlanned(NativeQueueIngress& input, NativeQueueSource& provider,
    const NativeQueueBatch& batch, NativeDispositionSchedule schedule, std::string& error) noexcept {
    return BeginInternal(input, provider, batch, true, schedule, error);
}
bool NativeEventDispositionLedger::BeginInternal(NativeQueueIngress& input, NativeQueueSource& provider,
    const NativeQueueBatch& batch, bool sinkPlan, NativeDispositionSchedule schedule, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (phase != Phase::Empty || !identity) return Fail(error, "Native disposition is not a fresh ledger");
    if (schedule != NativeDispositionSchedule::BeforeSessionDrain && schedule != NativeDispositionSchedule::AtMousePollEntry)
        return Fail(error, "Unsupported native disposition schedule");
    // Seal/input identity is supplied by actual NativeQueueIngress, not a caller
    // array of self-declared tags. All allocation occurs before the first effect.
    try {
        const auto& status = batch.Status();
        ingress = &input; source = &provider;
        context = {status.providerEpoch, status.generation, status.engineToken};
        receipt = {batch.Receipt(), identity, 1};
        if (batch.Events().empty() || batch.Events().size() > NativeQueueIngress::MaxEvents)
            return Fail(error, "Invalid or unavailable native disposition batch");
        // The sized constructor is allowed to throw even when MSVC debug
        // iterators allocate an empty container proxy. Its noexcept default
        // constructor would terminate before this operation can report OOM.
        std::vector<Record> candidate(0);
        candidate.reserve(batch.Events().size());
        for (const auto& event : batch.Events()) candidate.push_back({event.record, event.ignored});
        if (sinkPlan) emissions.reserve(MaxEmissions);
        // No reference into the caller's batch is used across Source::Observe.
        // Even releasing that owned object cannot replace the copied identity.
        if (!Check(error)) return false;
        passOrder = schedule == NativeDispositionSchedule::AtMousePollEntry ? PollFirstOrder : SessionFirstOrder;
        currentPass = passOrder[0];
        records.swap(candidate); planned = sinkPlan; phase = Phase::Between; error.clear(); return true;
    } catch (...) { return Fail(error, "Native disposition allocation failed"); }
}
bool NativeEventDispositionLedger::Matches(NativeDispositionRecord record) const noexcept {
    return record.receipt == receipt && next < records.size() && record.index == next &&
        record.queueSequence == records[next].tag.queue_sequence;
}
bool NativeEventDispositionLedger::TicketMatches(NativeDispositionTicket ticket) const noexcept {
    return Matches(ticket.record) && ticket.emission > first && ticket.emission <= issued &&
        ticket.emission == completed + 1;
}
void NativeEventDispositionLedger::Advance() noexcept {
    if (completed == issued && !inFlight) { ++next; phase = Phase::Between; }
}
bool NativeEventDispositionLedger::OpenRecord(std::size_t index, NativeDispositionRecord& out, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (phase != Phase::Between || index != next || next >= records.size())
        return Fail(error, "Native disposition record order mismatch");
    if (!Check(error)) return false;
    const NativeDispositionRecord result{receipt, records[next].tag.queue_sequence, next};
    first = issued; phase = Phase::Translating; out = result; error.clear(); return true;
}
bool NativeEventDispositionLedger::Issue(NativeDispositionRecord record, NativeDispositionTicket& out, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (planned || phase != Phase::Translating || !Matches(record) || records[next].ignored ||
        records[next].tag.kind == OQ4_QUEUE_SENTINEL || records[next].tag.kind == OQ4_QUEUE_FENCE || issued == MaxEmissions)
        return Fail(error, "Native disposition emission is invalid or exceeds budget");
    if (!Check(error)) return false;
    const NativeDispositionTicket result{record, issued + 1};
    ++issued; out = result; error.clear(); return true;
}
bool NativeEventDispositionLedger::PlannedTicketMatches(NativeDispositionTicket ticket) const noexcept {
    if (!planned || ticket.record.receipt != receipt || !ticket.emission || ticket.emission > emissions.size()) return false;
    const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
    return ticket.record.index == emission.record && emission.record < records.size() &&
        ticket.record.queueSequence == records[emission.record].tag.queue_sequence;
}
std::size_t NativeEventDispositionLedger::NextEmission(std::size_t cursor, NativeDispositionPass pass) const noexcept {
    while (cursor < emissions.size() && emissions[cursor].pass != pass) ++cursor;
    return cursor;
}
bool NativeEventDispositionLedger::Issue(NativeDispositionRecord record, NativeEmissionPlan plan,
    NativeDispositionTicket& out, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    const auto pass = static_cast<unsigned>(plan.pass);
    if (!planned || phase != Phase::Translating || !Matches(record) || records[next].ignored ||
        records[next].tag.kind == OQ4_QUEUE_SENTINEL || records[next].tag.kind == OQ4_QUEUE_FENCE ||
        pass > static_cast<unsigned>(NativeDispositionPass::SessionDeferred) || emissions.size() == MaxEmissions)
        return Fail(error, "Invalid fixed-sink emission plan");
    std::uint64_t trigger = 0;
    if (plan.pass == NativeDispositionPass::SessionDeferred) {
        if (!PlannedTicketMatches(plan.trigger) || plan.trigger.record != record)
            return Fail(error, "Deferred Session emission requires its exact Keyboard parent");
        const auto& parent = emissions[static_cast<std::size_t>(plan.trigger.emission - 1)];
        if (parent.pass != NativeDispositionPass::KeyboardPoll || parent.child)
            return Fail(error, "Deferred Session parent is not an unused Keyboard ticket");
        trigger = plan.trigger.emission;
    } else if (plan.trigger != NativeDispositionTicket{}) return Fail(error, "Unexpected deferred emission parent");
    if (!Check(error)) return false;
    const NativeDispositionTicket result{record, issued + 1};
    emissions.push_back({next, plan.pass, trigger, 0, 0, false}); // Pre-reserved, trivial values only.
    if (trigger) emissions[static_cast<std::size_t>(trigger - 1)].child = result.emission;
    ++issued; out = result; error.clear(); return true;
}
bool NativeEventDispositionLedger::AdmissionMatches(NativeDispositionTicket ticket) const noexcept {
    if (!PlannedTicketMatches(ticket)) return false;
    const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
    const auto pass = static_cast<std::size_t>(emission.pass);
    const bool deferred = emission.pass == NativeDispositionPass::SessionDeferred;
    return !emission.admission && ticket.emission - 1 == NextEmission(admissionCursor[pass], emission.pass) &&
        ((!deferred && phase == Phase::Translating && emission.record == next) ||
         (deferred && phase == Phase::Delivering && translationClosed && currentPass == NativeDispositionPass::KeyboardPoll &&
          inFlight && plannedInFlight == emission.trigger));
}
bool NativeEventDispositionLedger::CanAdmit(NativeDispositionTicket ticket) const noexcept {
    return thread == std::this_thread::get_id() && !calling && AdmissionMatches(ticket);
}
bool NativeEventDispositionLedger::Admit(NativeDispositionTicket ticket, NativeDispositionAdmission& out, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (!AdmissionMatches(ticket)) return Fail(error, "Admission is duplicate, reordered or outside its exact source/Keyboard delivery");
    auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
    const auto pass = static_cast<std::size_t>(emission.pass);
    if (!Check(error)) return false;
    const NativeDispositionAdmission result{ticket, admissionSerial + 1};
    emission.admission = ++admissionSerial;
    admissionCursor[pass] = static_cast<std::size_t>(ticket.emission);
    out = result; error.clear(); return true;
}
bool NativeEventDispositionLedger::SealTranslation(std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (!planned || phase != Phase::Between || next != records.size() || translationClosed)
        return Fail(error, "Fixed-sink translation is incomplete or already sealed");
    for (const auto& emission : emissions)
        if (emission.pass != NativeDispositionPass::SessionDeferred && !emission.admission)
            return Fail(error, "Translation has an unadmitted emission");
    if (!Check(error)) return false;
    translationClosed = true; phase = Phase::Delivering; error.clear(); return true;
}
bool NativeEventDispositionLedger::CompletePass(NativeDispositionPass pass, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (!planned || phase != Phase::Delivering || !translationClosed || inFlight || pass != currentPass ||
        NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass) != emissions.size())
        return Fail(error, "Fixed-sink pass has not reached its exact terminal checkpoint");
    if (!Check(error)) return false;
    if (++checkpoint == passOrder.size()) { passesComplete = true; phase = Phase::Between; }
    else currentPass = passOrder[checkpoint];
    error.clear(); return true;
}
bool NativeEventDispositionLedger::SealRecord(NativeDispositionRecord record, NativeRecordDisposition disposition, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (phase != Phase::Translating || !Matches(record)) return Fail(error, "Invalid native record seal");
    const auto kind = records[next].tag.kind;
    bool valid = false;
    if (kind == OQ4_QUEUE_SENTINEL) valid = disposition == NativeRecordDisposition::Sentinel;
    else if (kind == OQ4_QUEUE_FENCE) valid = disposition == NativeRecordDisposition::Fence;
    else if (records[next].ignored) valid = disposition == NativeRecordDisposition::Ignored;
    else if (issued != first) valid = disposition == NativeRecordDisposition::Emitted;
    else valid = disposition == NativeRecordDisposition::Immediate || disposition == NativeRecordDisposition::Ignored;
    if (!valid) return Fail(error, "Native record disposition does not match its effects");
    if (planned) {
        for (auto index = first; index < issued; ++index) {
            const auto& emission = emissions[static_cast<std::size_t>(index)];
            if (emission.pass != NativeDispositionPass::SessionDeferred && !emission.admission)
                return Fail(error, "Source record has an unadmitted sibling emission");
        }
    }
    if (!Check(error)) return false;
    if (planned) { ++next; phase = Phase::Between; }
    else { phase = Phase::Delivering; Advance(); }
    error.clear(); return true;
}
bool NativeEventDispositionLedger::CanBeginDelivery(NativeDispositionTicket ticket) const noexcept {
    if (thread != std::this_thread::get_id() || calling || !planned || phase != Phase::Delivering ||
        inFlight || !PlannedTicketMatches(ticket)) return false;
    const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
    return translationClosed && emission.pass == currentPass && emission.admission && !emission.done &&
        ticket.emission - 1 == NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass);
}
bool NativeEventDispositionLedger::BeginDelivery(NativeDispositionTicket ticket, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (planned) {
        if (phase != Phase::Delivering || inFlight || !PlannedTicketMatches(ticket))
            return Fail(error, "Invalid fixed-sink delivery ticket");
        const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
        if (!translationClosed || emission.pass != currentPass || !emission.admission || emission.done ||
            ticket.emission - 1 != NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass))
            return Fail(error, "Fixed-sink delivery is unadmitted, duplicate or reordered");
        if (!Check(error)) return false;
        inFlight = true; plannedInFlight = ticket.emission; error.clear(); return true;
    }
    if (phase != Phase::Delivering || inFlight || !TicketMatches(ticket))
        return Fail(error, "Unknown, duplicate or reordered native delivery ticket");
    if (!Check(error)) return false;
    inFlight = true; error.clear(); return true;
}
bool NativeEventDispositionLedger::CompleteDelivery(NativeDispositionTicket ticket, NativeDeliveryDisposition disposition, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (planned) {
        if (phase != Phase::Delivering || !inFlight || plannedInFlight != ticket.emission || !PlannedTicketMatches(ticket) ||
            (disposition != NativeDeliveryDisposition::Delivered && disposition != NativeDeliveryDisposition::AuditedDiscard))
            return Fail(error, "Fixed-sink completion does not match its active ticket");
        auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
        if (emission.child && !emissions[static_cast<std::size_t>(emission.child - 1)].admission)
            return Fail(error, "Keyboard delivery did not admit its reserved Session event");
        if (!Check(error)) return false;
        emission.done = true; deliveryCursor[static_cast<std::size_t>(currentPass)] = static_cast<std::size_t>(ticket.emission);
        ++completed; inFlight = false; plannedInFlight = 0; error.clear(); return true;
    }
    if (phase != Phase::Delivering || !inFlight || !TicketMatches(ticket) ||
        (disposition != NativeDeliveryDisposition::Delivered && disposition != NativeDeliveryDisposition::AuditedDiscard))
        return Fail(error, "Native delivery did not complete its exact active ticket");
    if (!Check(error)) return false;
    ++completed; inFlight = false; Advance(); error.clear(); return true;
}
bool NativeEventDispositionLedger::Seal(NativeDispositionReceipt& out, std::string& error) noexcept {
    if (!Enter(error)) return false;
    Guard guard{*this};
    if (phase != Phase::Between || next != records.size() || inFlight || completed != issued || (planned && !passesComplete))
        return Fail(error, "Native disposition batch is incomplete");
    if (!Check(error)) return false;
    phase = Phase::Complete; out = receipt; error.clear(); return true;
}
bool NativeEventDispositionLedger::Current(NativeDispositionReceipt expected) const noexcept {
    return thread == std::this_thread::get_id() && !calling && phase == Phase::Complete &&
        expected == receipt && probe.Current(context);
}
bool NativeEventDispositionLedger::Retire() noexcept {
    if (thread != std::this_thread::get_id()) return false;
    phase = Phase::Retired; return true;
}
bool NativeEventDispositionLedger::NeedsRetirement() const noexcept {
    return thread == std::this_thread::get_id() && phase == Phase::Retired;
}
bool NativeEventDispositionLedger::InspectIssuedForRetirement(NativeDispositionTicket ticket,
    NativeIssuedEmission& out) const noexcept {
    if (thread != std::this_thread::get_id() || calling || phase != Phase::Retired || !PlannedTicketMatches(ticket)) return false;
    const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];
    NativeIssuedEmission result;
    result.ticket = ticket;
    result.pass = emission.pass;
    result.admissionSerial = emission.admission;
    result.terminal = emission.done;
    result.inFlight = inFlight && plannedInFlight == ticket.emission;
    if (emission.trigger) result.trigger = {ticket.record, emission.trigger};
    out = result;
    return true;
}
bool NativeEventDispositionLedger::QueryRetirement(NativeDispositionRetirement& out) const noexcept {
    if (thread != std::this_thread::get_id() || calling || phase != Phase::Retired || !planned ||
        !receipt.ledger || !receipt.serial || issued != emissions.size()) return false;
    out = {receipt, issued, inFlight ? plannedInFlight : 0};
    return true;
}
} // namespace openq4

File: tools/tests/native_event_disposition.py
#!/usr/bin/env python3
"""Actual portable ordinary-disposition ledger, ingress and continuity model.

Uses pinned SDL public types; source Poll/Observe/Copy and engine delivery are
counted doubles. No SDL pump, queue-sidecar integration, native/GUI activation,
OS input, full engine build, journal change or game is performed.
"""
from pathlib import Path
import argparse
import configparser
import hashlib
import importlib.util
import json
import os
import re
import subprocess
import tempfile

ROOT = Path(__file__).resolve().parents[2]
SDL_HEADERS = ("SDL_atomic.h SDL_audio.h SDL_begin_code.h SDL_blendmode.h SDL_camera.h "
    "SDL_close_code.h SDL_endian.h SDL_error.h SDL_events.h SDL_gamepad.h SDL_guid.h "
    "SDL_init.h SDL_iostream.h SDL_joystick.h SDL_keyboard.h SDL_keycode.h SDL_mouse.h "
    "SDL_mutex.h SDL_pen.h SDL_pixels.h SDL_platform_defines.h SDL_power.h SDL_properties.h "
    "SDL_rect.h SDL_scancode.h SDL_sensor.h SDL_stdinc.h SDL_surface.h SDL_thread.h "
    "SDL_touch.h SDL_video.h").split()

def sha(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()

def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--sdl-source', type=Path)
    parser.add_argument('--compiler', default='clang++')
    parser.add_argument('--msvc-debug', action='store_true')
    parser.add_argument('--sanitizers', action='store_true')
    parser.add_argument('--no-mutations', action='store_true')
    parser.add_argument('--test-source', default='tools/tests/native/NativeRouteAccountingTest.cpp')
    args = parser.parse_args()
    (ROOT/'.tmp').mkdir(exist_ok=True)
    folder = Path(tempfile.mkdtemp(prefix='native-event-disposition-', dir=ROOT/'.tmp'))
    helper_path = ROOT/'tools/tests/sdl3_clipboard_status.py'
    spec = importlib.util.spec_from_file_location('sdl_source_helper', helper_path)
    helper = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(helper)
    helper.ROOT = ROOT
    helper.FILES = ['include/SDL3/'+name for name in SDL_HEADERS]
    config = configparser.ConfigParser()
    config.read(ROOT/'subprojects/sdl3.wrap', encoding='utf-8')
    specification = dict(config['wrap-file'])
    assert specification['directory'] == 'SDL3-3.4.10', 'Review SDL public header closure on dependency update'
    sdl, provision, _ = helper.provision_source(args.sdl_source, specification, folder)
    names = ['src/sys/sdl3/NativeEventDisposition.h', 'src/sys/sdl3/NativeEventDisposition.cpp',
             'src/sys/sdl3/NativeQueueBatch.h', 'src/sys/sdl3/NativeQueueBatch.cpp',
             'src/ui/retained/TextInput.h', 'src/ui/retained/TextInput.cpp',
             'src/sys/EventQueueContinuity.h', 'tools/tests/native/NativeEventDispositionTest.cpp',
             'subprojects/packagefiles/sdl3/include/SDL3/SDL_openq4_native_fence.h',
             'tools/tests/native_event_disposition.py', 'tools/tests/sdl3_clipboard_status.py', 'subprojects/sdl3.wrap',
             'tools/tests/native/NativeRouteAccountingTest.cpp', args.test_source]
    before = {name: sha(ROOT/name) for name in names}
    flags = [args.compiler, '-std=c++20', '-Wall', '-Wextra', '-Werror', '-DSDL_STATIC_LIB']
    if os.name != 'nt': flags += ['-pthread']
    include = [ROOT, ROOT/'subprojects/packagefiles/sdl3/include', sdl/'include']
    flags += ['-I'+str(path) for path in include]
    if args.sanitizers:
        flags += ['-fsanitize=address,undefined', '-fno-omit-frame-pointer', '-g0']
    if args.msvc_debug:
        assert not args.sanitizers
        flags = [args.compiler, '/nologo', '/std:c++20', '/EHsc', '/W3', '/WX', '/MTd',
                 '/D_ITERATOR_DEBUG_LEVEL=2', '/DSDL_STATIC_LIB'] + ['/I'+str(path) for path in include]
    env = {**os.environ, 'TEMP': str(folder), 'TMP': str(folder), 'TMPDIR': str(folder)}
    report = {'passed': False, 'scope': __doc__, 'sources': before, 'cases': [], 'source_provision': provision,
              'sdl_headers': {name: sha(sdl/'include/SDL3'/name) for name in SDL_HEADERS}}
    source = ROOT/'src/sys/sdl3/NativeEventDisposition.cpp'
    text = source.read_text(encoding='utf-8-sig')
    projection = folder/'NativeEventDisposition.cpp'
    batch_source = ROOT/'src/sys/sdl3/NativeQueueBatch.cpp'
    batch_text = batch_source.read_text(encoding='utf-8-sig')
    batch_projection = folder/'NativeQueueBatch.cpp'
    batch_projection.write_text(batch_text, encoding='utf-8', newline='\n')
    # The quoted include in this copied TU resolves to its unchanged real header.
    flags += (['/I'+str(source.parent)] if args.msvc_debug else ['-I'+str(source.parent)])
    projection.write_text(text, encoding='utf-8', newline='\n')
    def run(command, name):
        result = subprocess.run(command, cwd=folder, env=env, capture_output=True, text=True,
                                encoding='utf-8', errors='replace', timeout=120)
        log = folder/(name+'.log')
        log.write_text(result.stdout+result.stderr, encoding='utf-8')
        return result, log
    def compile(path, name):
        obj = folder/(name+'.obj')
        command = flags + (['/c', str(path), '/Fo:'+str(obj)] if args.msvc_debug else ['-c', str(path), '-o', str(obj)])
        result, log = run(command, name+'-compile')
        if result.returncode:
            raise RuntimeError(str(log)+'\n'+result.stdout+result.stderr)
        return obj
    try:
        common = [compile(ROOT/name, 'common-'+str(n)) for n, name in enumerate([
            'src/ui/retained/TextInput.cpp', args.test_source])]
        batch_base = compile(batch_projection, 'batch-base')
        ledger_base = None
        def case(name, mutation=False, batch_mutation=False):
            nonlocal ledger_base
            obj = compile(batch_projection if batch_mutation else projection, name)
            if not mutation: ledger_base = obj
            binary = folder/(name+'.exe')
            command = flags + [str(path) for path in [*common, obj, ledger_base if batch_mutation else batch_base]]
            command += ['/Fe:'+str(binary)] if args.msvc_debug else ['-o', str(binary)]
            result, log = run(command, name+'-link')
            if result.returncode: raise RuntimeError(str(log)+'\n'+result.stdout+result.stderr)
            result, log = run([str(binary)], name+'-run')
            record = {'name': name, 'exit': result.returncode, 'source_sha256': sha(projection),
                      'batch_source_sha256': sha(batch_projection),
                      'binary_sha256': sha(binary), 'log_sha256': sha(log)}
            report['cases'].append(record)
            if mutation:
                if not result.returncode or 'FAIL' not in result.stdout+result.stderr:
                    raise RuntimeError('Mutation not assertion-rejected: '+name+'\n'+result.stdout+result.stderr)
                record['rejected'] = True
            else:
                if result.returncode: raise RuntimeError(str(log)+'\n'+result.stdout+result.stderr)
                record['checks'] = int(re.findall(r'PASS (\d+) checks', result.stdout)[-1])
                print(result.stdout.strip(), flush=True)
        case('baseline')
        mutations = {
            'skip-record-order': ('index != next ||', '(void(index),false) ||'),
            'skip-record-identity': ('record.receipt == receipt &&', 'true &&'),
            'skip-record-sequence': ('record.queueSequence == records[next].tag.queue_sequence', 'true'),
            'skip-ticket-order': ('ticket.emission == completed + 1', 'true'),
            'skip-delivery-entry': ('phase != Phase::Delivering || !inFlight || !TicketMatches(ticket)', 'phase != Phase::Delivering || !TicketMatches(ticket)'),
            'allow-double-delivery-entry': ('phase != Phase::Delivering || inFlight || !TicketMatches(ticket)', 'phase != Phase::Delivering || !TicketMatches(ticket)'),
            'advance-before-delivered': ('if (completed == issued && !inFlight)', 'if (!inFlight)'),
            'skip-probe-after-source': ('!probe.Current(context)) return Fail', 'false) return Fail'),
            'skip-ingress-validation': ('!ingress->Validate(*source, receipt.batch, error) ||', 'false ||'),
            'skip-current-continuity': ('expected == receipt && probe.Current(context)', 'expected == receipt'),
            'ignore-reentry-latch': ('if (calling) return Fail(error, "Reentrant native disposition requires retirement");', 'if (calling) return false;'),
            'skip-postcallback-retirement': ('receipt.batch, error) || phase == Phase::Retired ||', 'receipt.batch, error) ||'),
            'permit-reused-ledger': ('phase != Phase::Empty || !identity', '!identity'),
            'skip-emission-cap': ('issued == MaxEmissions', 'false'),
            'allow-fence-emission': ('records[next].tag.kind == OQ4_QUEUE_FENCE || issued == MaxEmissions', 'issued == MaxEmissions'),
            'skip-batch-completion': ('phase != Phase::Between || next != records.size() || inFlight || completed != issued', 'phase == Phase::Retired'),
        }
        planned_mutations = {
            'retired-inspection-before-retire': ('calling || phase != Phase::Retired || !PlannedTicketMatches(ticket)', 'calling || !PlannedTicketMatches(ticket)'),
            'retired-inspection-reentry': ('calling || phase != Phase::Retired || !PlannedTicketMatches(ticket)', 'phase != Phase::Retired || !PlannedTicketMatches(ticket)'),
            'retired-inspection-foreign-thread': ('if (thread != std::this_thread::get_id() || calling || phase != Phase::Retired || !PlannedTicketMatches(ticket))', 'if (calling || phase != Phase::Retired || !PlannedTicketMatches(ticket))'),
            'retired-inspection-requires-admission': ('const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];\n    NativeIssuedEmission result;', 'const auto& emission = emissions[static_cast<std::size_t>(ticket.emission - 1)];\n    if (!emission.admission) return false;\n    NativeIssuedEmission result;'),
            'retired-inspection-fabricates-terminal': ('result.terminal = emission.done;', 'result.terminal = true;'),
            'retired-inspection-loses-inflight': ('result.inFlight = inFlight && plannedInFlight == ticket.emission;', 'result.inFlight = false;'),
            'retired-inspection-loses-trigger': ('if (emission.trigger) result.trigger = {ticket.record, emission.trigger};', 'if (false) result.trigger = {ticket.record, emission.trigger};'),
            'schedule-default-changed': ('return BeginPlanned(input, provider, batch, NativeDispositionSchedule::BeforeSessionDrain, error);', 'return BeginPlanned(input, provider, batch, NativeDispositionSchedule::AtMousePollEntry, error);'),
            'schedule-unknown-accepted': ('if (schedule != NativeDispositionSchedule::BeforeSessionDrain && schedule != NativeDispositionSchedule::AtMousePollEntry)', 'if (false)'),
            'schedule-selection-swapped': ('? PollFirstOrder : SessionFirstOrder', '? SessionFirstOrder : PollFirstOrder'),
            'schedule-enum-increment': ('else currentPass = passOrder[checkpoint];', 'else currentPass = static_cast<NativeDispositionPass>(static_cast<unsigned>(currentPass) + 1);'),
            'schedule-empty-pass-reordered': ('inFlight || pass != currentPass ||', 'inFlight || (void(pass),false) ||'),
            'schedule-skip-final-pass': ('++checkpoint == passOrder.size()', '++checkpoint == passOrder.size()-1'),
            'plan-skip-admission-order': ('ticket.emission - 1 == NextEmission(admissionCursor[pass], emission.pass)', '(void(pass),true)'),
            'plan-skip-delivery-order': ('ticket.emission - 1 != NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass)', 'false'),
            'plan-skip-pass-order': ('!translationClosed || emission.pass != currentPass || !emission.admission || emission.done ||\n            ticket.emission - 1 != NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass)', '!translationClosed || !emission.admission || emission.done'),
            'plan-skip-child-parent': ('inFlight && plannedInFlight == emission.trigger', 'true'),
            'plan-skip-parent-child-admit': ('if (emission.child && !emissions[static_cast<std::size_t>(emission.child - 1)].admission)', 'if (false)'),
            'plan-skip-parent-kind': ('parent.pass != NativeDispositionPass::KeyboardPoll || parent.child', 'parent.child'),
            'plan-skip-parent-reuse': ('parent.pass != NativeDispositionPass::KeyboardPoll || parent.child', 'parent.pass != NativeDispositionPass::KeyboardPoll'),
            'plan-skip-cap': ('emissions.size() == MaxEmissions', 'false'),
            'plan-skip-ticket-ledger': ('ticket.record.receipt != receipt ||', ''),
            'plan-skip-ticket-sequence': ('ticket.record.queueSequence == records[emission.record].tag.queue_sequence', 'true'),
            'plan-skip-sibling-admission': ('if (emission.pass != NativeDispositionPass::SessionDeferred && !emission.admission)\n                return Fail(error, "Source record has an unadmitted sibling emission");', 'if (false && emission.admission) return false;'),
            'plan-skip-pass-drain': ('NextEmission(deliveryCursor[static_cast<std::size_t>(currentPass)], currentPass) != emissions.size()', 'false'),
            'plan-skip-all-passes': (' || (planned && !passesComplete)', ''),
        }
        quarantine_mutations = {
            'quarantine-lose-prefix': ('owner.quarantine=std::move(candidate);', 'candidate.reset();'),
            'quarantine-lose-current-copy': ('provisional=false;\n            failure.copied=true;', 'provisional=true;\n            failure.copied=true;'),
            'quarantine-keep-rejected-slot': ('if (candidate && provisional) candidate->prefix->events.pop_back();', 'if (false && candidate && provisional) candidate->prefix->events.pop_back();'),
            'quarantine-reuse-oom-generation': ('lastEpoch=status.providerEpoch; lastGeneration=status.generation; lastSequence=sequence;', 'lastSequence=sequence;'),
            'quarantine-reset-with-owned-prefix': ('if (quarantine) return Fail', 'if (false) return Fail'),
            'quarantine-wrong-thread-take': ('if (std::this_thread::get_id()!=thread) return false;\n    if (calling)', 'if (calling)'),
            'quarantine-hide-removal-boundary': ('failure.returnedOne=true;', 'failure.returnedOne=false;'),
            'quarantine-forget-copied-boundary': ('failure.copied=true; failure.reason=', 'failure.copied=false; failure.reason='),
        }
        if Path(args.test_source).name == 'NativeRouteAccountingTest.cpp':
            mutations.update(planned_mutations)
        if not args.no_mutations:
            for name, (old, new) in mutations.items():
                assert text.count(old) == 1, (name, 'mutation anchor not unique')
                projection.write_text(text.replace(old, new), encoding='utf-8', newline='\n')
                case(name, True)
            projection.write_text(text, encoding='utf-8', newline='\n')
            if Path(args.test_source).name == 'NativeRouteAccountingTest.cpp':
                for name, (old, new) in quarantine_mutations.items():
                    assert batch_text.count(old) == 1, (name, 'mutation anchor not unique')
                    batch_projection.write_text(batch_text.replace(old, new), encoding='utf-8', newline='\n')
                    case(name, True, True)
        report['sources_unchanged'] = before == {name: sha(ROOT/name) for name in names}
        report['passed'] = report['sources_unchanged']
    finally:
        projection.write_text(text, encoding='utf-8', newline='\n')
        batch_projection.write_text(batch_text, encoding='utf-8', newline='\n')
        report['logs'] = {p.name: sha(p) for p in folder.glob('*.log')}
        report['objects'] = {p.name: sha(p) for p in folder.glob('*.obj')}
        (folder/'result.json').write_text(json.dumps(report, indent=2)+'\n', encoding='utf-8')
        print(folder/'result.json', flush=True)
    if not report['passed']: raise SystemExit('Source binding failed')

if __name__ == '__main__':
    main()
